oak_restricted_kernel/src/ghcb.rs

```diff
@@ -0,0 +1,187 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::mm::{ENCRYPTED_BIT_POSITION, KERNEL_OFFSET};
+use sev_guest::{
+    ghcb::{Ghcb, GhcbProtocol},
+    msr::{
+        change_snp_page_state, register_ghcb_location, PageAssignment, RegisterGhcbGpaRequest,
+        SnpPageStateChangeRequest,
+    },
+    Translator,
+};
+use x86_64::{
+    addr::{PhysAddr, VirtAddr},
+    align_down,
+    instructions::tlb::flush_all,
+    registers::control::Cr3,
+    structures::paging::{
+        page_table::{PageTable, PageTableFlags},
+        PageSize, Size1GiB, Size2MiB, Size4KiB,
+    },
+};
+
+/// The mask for the encrypted bit
+const ENCRYPTED_BIT: u64 = 1 << ENCRYPTED_BIT_POSITION;
+
+/// A wrapper to ensure that the GHCB is alone in a 2MiB page.
+///
+/// We use 2MiB pages during early boot, so we must make sure there are no other kernel data
+/// structures located in the page so that we can safely share the page with the hypervisor without
+/// leaking other data.
+#[derive(Debug)]
+#[repr(C, align(2097152))]
+struct GhcbAlignmentWrapper {
+    ghcb: Ghcb,
+}
+
+static_assertions::assert_eq_size!(GhcbAlignmentWrapper, [u8; Size2MiB::SIZE as usize]);
+
+static mut GHCB_WRAPPER: GhcbAlignmentWrapper = GhcbAlignmentWrapper { ghcb: Ghcb::new() };
+
+/// Initializes the GHCB and shares it with the hypervisor during early boot.
+pub fn init_ghcb_early(snp_enabled: bool) -> GhcbProtocol<'static, Ghcb> {
+    // Safety: This is called only during early boot, so there is only a single execution context.
+    let ghcb = unsafe { &mut GHCB_WRAPPER.ghcb };
+    let translate = |vaddr: VirtAddr| {
+        let address = vaddr.as_u64();
+        // We assume a fixed offset mapping for kernel pages.
+        if address < KERNEL_OFFSET {
+            None
+        } else {
+            Some(PhysAddr::new(address - KERNEL_OFFSET))
+        }
+    };
+    share_ghcb_with_hypervisor(ghcb, snp_enabled, translate);
+    ghcb.reset();
+
+    GhcbProtocol::new(ghcb, translate)
+}
+
+/// Marks the page containing the GHCB data structure to be shared with the hypervisor.
+fn share_ghcb_with_hypervisor<VP: Translator>(ghcb: &Ghcb, snp_enabled: bool, translate: VP) {
+    let ghcb_address = VirtAddr::new(ghcb as *const Ghcb as usize as u64);
```

Comment:
Use `VirtAddr::from_ptr()`, that saves two casts here.

Reply:
Done